File: conformance/node/tests/twilio.test.ts
// twilio-node conformance: a custom httpClient (the SDK's documented
// injection point) rewrites scheme+host to the adapter — Twilio signs no
// request URLs, so this is sound. Lifecycle via SDK fetches, and callback
// verification through twilio-node's own validateRequest.
import { describe, expect, test } from "bun:test";
import twilio from "twilio";
import { bootAdapter, type Harness } from "../helpers";

const SID = "AC0123456789abcdef0123456789abcdef".replace("AC0", "AC" + "0");
const AUTH_TOKEN = "feed0000face1111beef2222cafe3333";

// FetchBackedClient implements the RequestClient contract
// ({statusCode, body, headers}) against the stunt base.
class FetchBackedClient {
  constructor(private base: string) {}
  async request(opts: Record<string, any>) {
    const u = new URL(opts.uri);
    const url = new URL(u.pathname + u.search, this.base);
    if (opts.params) {
      for (const [k, v] of Object.entries(opts.params)) url.searchParams.set(k, String(v));
    }
    const headers: Record<string, string> = {};
    for (const [k, v] of Object.entries(opts.headers ?? {})) if (v != null) headers[k] = String(v);
    if (opts.username && opts.password) {
      headers.Authorization =
        "Basic " + Buffer.from(`${opts.username}:${opts.password}`).toString("base64");
    }
    let body: string | undefined;
    const method = String(opts.method ?? "get").toUpperCase();
    if (opts.data && headers["Content-Type"] === "application/x-www-form-urlencoded") {
      body = new URLSearchParams(
        Object.entries(opts.data).map(([k, v]) => [k, String(v)]),
      ).toString();
    } else if (opts.data != null) {
      body = JSON.stringify(opts.data);
    }
    const res = await fetch(url, { method, headers, body });
    const text = await res.text();
    let parsed: unknown = text;
    try {
      parsed = JSON.parse(text);
    } catch {}
    const outHeaders: Record<string, string> = {};
    res.headers.forEach((v, k) => (outHeaders[k] = v));
    return { statusCode: res.status, body: parsed, headers: outHeaders };
  }
}

async function pollStatus(h: Harness, sid: string, want: string[]): Promise<string> {
  const deadline = Date.now() + 10_000;
  let status = "";
  while (Date.now() < deadline) {
    const res = await fetch(`${h.base}/2010-04-01/Accounts/${SID}/Messages/${sid}.json`, {
      headers: { Authorization: "Basic " + Buffer.from(`${SID}:${AUTH_TOKEN}`).toString("base64") },
    });
    const msg = (await res.json()) as Record<string, any>;
    if (want.includes(msg.status)) {
      status = msg.status;
      break;
    }
    await Bun.sleep(300);
  }
  return status;
}

describe("twilio-node against twilio-style", () => {
  test(
    "lifecycle + signed callbacks via validateRequest",
    async () => {
      const h = await bootAdapter("twilio-style");
      try {
        const client = twilio(SID, AUTH_TOKEN, {
          httpClient: new FetchBackedClient(h.base) as any,
          lazyLoading: false,
        });

        // ===== Create + SDK fetches drive the lifecycle =====
        const msg = await client.api.v2010.accounts(SID).messages.create({
          to: "+15550002222",
          from: "+15550001111",
          body: "node conformance hello",
        });
        expect(msg.sid).toMatch(/^SM/);

        // Poll via the SDK until terminal (drives derive-on-read
        // transitions, which fire the callbacks).
        const deadline = Date.now() + 10_000;
        let status = "";
        // Drive to DELIVERED (not just sent): the delivered callback only
        // fires on the read that derives the second transition.
        while (Date.now() < deadline) {
          const m = await client.api.v2010.accounts(SID).messages(msg.sid).fetch();
          if (["delivered", "failed"].includes(m.status)) {
            status = m.status;
            break;
          }
          await Bun.sleep(300);
        }
        expect(status).toBe("delivered");

        // ===== Callbacks verified by twilio-node's own validateRequest =====
        const first = await h.waitFor((d) => d.headers["x-twilio-signature"] != null);
        const params = Object.fromEntries(new URLSearchParams(first.body));
        const url = first.url;
        const ok = twilio.validateRequest(AUTH_TOKEN, first.headers["x-twilio-signature"], url, params);
        expect(ok).toBe(true);

        // The terminal (delivered) callback MUST arrive — the loop only
        // verifies what's present, so a dropped unsigned second callback
        // must fail here, not slip through.
        await h.waitFor((d) => d !== first && d.headers["x-twilio-signature"] != null, 8_000);
        for (const d of h.deliveries()) {
          if (!d.headers["x-twilio-signature"]) continue;
          const p = Object.fromEntries(new URLSearchParams(d.body));
          expect(twilio.validateRequest(AUTH_TOKEN, d.headers["x-twilio-signature"], d.url, p)).toBe(
            true,
          );
        }

        // ===== The magic invalid number -> failed =====
        const bad = await client.api.v2010.accounts(SID).messages.create({
          to: "+15005550001",
          from: "+15550001111",
          body: "should fail",
        });
        const badStatus = await pollStatus(h, bad.sid, ["failed", "sent"]);
        expect(badStatus).toBe("failed");
      } finally {
        await h.stop();
      }
    },
    { timeout: 120_000 },
  );
});

File: conformance/twilio_test.go
package conformance

import (
	"crypto/hmac"
	"crypto/sha1"
	"encoding/base64"
	"io"
	"net/http"
	"net/http/httptest"
	"net/url"
	"sort"
	"strings"
	"sync"
	"testing"
	"time"

	"github.com/twilio/twilio-go"
	tclient "github.com/twilio/twilio-go/client"
	v2010 "github.com/twilio/twilio-go/rest/api/v2010"
)

// twilioMockAuthToken mirrors the adapter's documented mock auth token —
// real-looking 32-hex, because official SDKs (twilio-go) client-side
// validate credentials as alphanumeric and reject underscore tokens.
const twilioMockAuthToken = "feed0000face1111beef2222cafe3333"

// TestTwilioSDKConformance drives the official twilio-go SDK against the
// twilio-style adapter with the documented test credentials: message
// create, the queued->sent->delivered lifecycle through SDK Fetch polling,
// the magic invalid-number failure trigger, list filters, and a signed
// status-callback webhook verified with Twilio's documented formula.
func TestTwilioSDKConformance(t *testing.T) {
	base := Boot(t, "twilio-style")

	const sid = "AC" + "0123456789abcdef0123456789abcdef"
	client := newTwilioClient(t, base, sid)

	// ===== Create + lifecycle (queued -> sent -> delivered via SDK fetch) =====

	msg, err := client.Api.CreateMessage(&v2010.CreateMessageParams{
		To:   ptr("+15550002222"),
		From: ptr("+15550001111"),
		Body: ptr("conformance hello"),
	})
	if err != nil {
		t.Fatalf("CreateMessage: %v", err)
	}
	msgSid := deref(msg.Sid)
	if !strings.HasPrefix(msgSid, "SM") {
		t.Fatalf("message Sid = %q, want SM prefix", msgSid)
	}
	Record(t, "twilio-go", "twilio-style", "CreateMessage (SM sid assigned)")

	var terminal string
	deadline := time.Now().Add(10 * time.Second)
	for time.Now().Before(deadline) {
		m, err := client.Api.FetchMessage(msgSid, nil)
		if err != nil {
			t.Fatalf("FetchMessage: %v", err)
		}
		if st := deref(m.Status); st == "delivered" || st == "failed" || st == "sent" {
			terminal = st
			break
		}
		time.Sleep(300 * time.Millisecond)
	}
	if terminal != "delivered" && terminal != "sent" {
		t.Fatalf("lifecycle poll ended at %q, want sent/delivered (derive-on-read transitions)", terminal)
	}
	Record(t, "twilio-go", "twilio-style", "FetchMessage poll reaches terminal state")

	// ===== The magic invalid number -> failed =====

	bad, err := client.Api.CreateMessage(&v2010.CreateMessageParams{
		To:   ptr("+15005550001"),
		From: ptr("+15550001111"),
		Body: ptr("should fail"),
	})
	if err != nil {
		t.Fatalf("CreateMessage invalid number: %v", err)
	}
	badSid := deref(bad.Sid)
	badTerminal := ""
	deadline = time.Now().Add(10 * time.Second)
	for time.Now().Before(deadline) {
		m, err := client.Api.FetchMessage(badSid, nil)
		if err != nil {
			t.Fatalf("FetchMessage bad: %v", err)
		}
		if st := deref(m.Status); st != "queued" && st != "" {
			badTerminal = st
			break
		}
		time.Sleep(300 * time.Millisecond)
	}
	if badTerminal != "failed" {
		t.Fatalf("invalid number terminal status = %q, want failed", badTerminal)
	}
	Record(t, "twilio-go", "twilio-style", "+15005550001 magic number -> failed")

	// ===== List with a To filter (only the matching message) =====

	list, err := client.Api.ListMessage(&v2010.ListMessageParams{
		To: ptr("+15550002222"),
	})
	if err != nil {
		t.Fatalf("ListMessage: %v", err)
	}
	for _, m := range list {
		if deref(m.To) != "+15550002222" {
			t.Fatalf("filter leaked: message To=%q in a To=+15550002222 list", deref(m.To))
		}
	}
	if len(list) == 0 {
		t.Fatal("To filter returned nothing")
	}
	Record(t, "twilio-go", "twilio-style", "ListMessage To filter honored")

	// ===== Signed status-callback webhook (Twilio's documented formula) =====

	var mu sync.Mutex
	var callbacks []struct {
		url     string
		body    string
		headers http.Header
	}
	sink := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		b, _ := io.ReadAll(r.Body)
		mu.Lock()
		callbacks = append(callbacks, struct {
			url     string
			body    string
			headers http.Header
		}{"http://" + r.Host + r.URL.RequestURI(), string(b), r.Header.Clone()})
		mu.Unlock()
		w.WriteHeader(200)
	}))
	defer sink.Close()

	// The adapter signs deliveries against the engine-registered sink.
	cbBase := Boot(t, "twilio-style", sink.URL)
	cbClient := newTwilioClient(t, cbBase, sid)
	cbMsg, err := cbClient.Api.CreateMessage(&v2010.CreateMessageParams{
		To:   ptr("+15550003333"),
		From: ptr("+15550001111"),
		Body: ptr("with callback"),
	})
	if err != nil {
		t.Fatalf("CreateMessage callback: %v", err)
	}
	// Callbacks fire per status TRANSITION and transitions are read-driven
	// (derive-on-read) — poll this engine's message to delivered so BOTH
	// the sent and delivered callbacks fire.
	cbSid := deref(cbMsg.Sid)
	deadline = time.Now().Add(10 * time.Second)
	for time.Now().Before(deadline) {
		m, err := cbClient.Api.FetchMessage(cbSid, nil)
		if err != nil {
			t.Fatalf("FetchMessage cb: %v", err)
		}
		if st := deref(m.Status); st == "delivered" || st == "failed" {
			break
		}
		time.Sleep(300 * time.Millisecond)
	}

	deadline = time.Now().Add(10 * time.Second)
	for time.Now().Before(deadline) {
		mu.Lock()
		n := len(callbacks)
		mu.Unlock()
		if n >= 2 { // sent + delivered hops
			break
		}
		time.Sleep(200 * time.Millisecond)
	}
	mu.Lock()
	defer mu.Unlock()
	if len(callbacks) == 0 {
		t.Fatal("no status callbacks arrived at the sink")
	}
	for _, cb := range callbacks {
		sig := cb.headers.Get("X-Twilio-Signature")
		if sig == "" {
			t.Fatal("callback without X-Twilio-Signature")
		}
		// base64(HMAC-SHA1(authToken, url + sorted form params)).
		vals, err := url.ParseQuery(cb.body)
		if err != nil {
			t.Fatalf("callback body not a form: %v", err)
		}
		keys := make([]string, 0, len(vals))
		for k := range vals {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		var sb strings.Builder
		sb.WriteString(cb.url)
		for _, k := range keys {
			sb.WriteString(k + vals.Get(k))
		}
		mac := hmac.New(sha1.New, []byte(twilioMockAuthToken))
		mac.Write([]byte(sb.String()))
		want := base64.StdEncoding.EncodeToString(mac.Sum(nil))
		if sig != want {
			t.Fatalf("X-Twilio-Signature mismatch:\n got %q\nwant %q (url %s)", sig, want, cb.url)
		}
	}
	Record(t, "twilio-go", "twilio-style", "status callbacks verify against Twilio's HMAC-SHA1 formula")
}

// newTwilioClient builds the official client on the rewrite transport —
// the SDK hardcodes api.twilio.com, and accepts a custom BaseClient.
func newTwilioClient(t *testing.T, base, sid string) *twilio.RestClient {
	t.Helper()
	bc := &tclient.Client{
		Credentials: tclient.NewCredentials(sid, twilioMockAuthToken),
		HTTPClient:  RewriteClient(t, base),
	}
	bc.SetAccountSid(sid)
	return twilio.NewRestClientWithParams(twilio.ClientParams{Client: bc})
}

func deref[T any](p *T) T {
	if p == nil {
		var zero T
		return zero
	}
	return *p
}
